truffle/com.oracle.truffle.api/src/com/oracle/truffle/api/nodes/LoopNode.java

```diff
@@ -167,8 +172,91 @@ protected LoopNode() {
      */
     public static void reportLoopCount(Node source, int iterations) {
         if (CompilerDirectives.inInterpreter()) {
-            ACCESSOR.onLoopCount(source, iterations);
+            if (SUPPORT == null) {
+                final TruffleRuntime runtime = Truffle.getRuntime();
+                SUPPORT = runtime.getCapability(TVMCI.class);
+                if (SUPPORT == null) {
+                    SUPPORT = new NoOp();
+                }
+            }
+            SUPPORT.onLoopCount(source, iterations);
+        }
+    }
+
+    /**
+     * An interface between Truffle API and hosting virtual machine. Not interesting for regular
+     * Truffle API users. Acronym for Truffle Virtual Machine Compiler Interface.
+     * 
+     * @since 0.12
+     */
+    protected abstract class TVMCI {
+        /** Only useful for virtual machine implementors. 
+         * @since 0.12 
+         */
+        protected TVMCI() {
+        }
+
+        // the class is a member class to not be (easily) visible in Javadoc
+
+        /**
+         * Reports the execution count of a loop.
+         *
+         * @param source the Node which invoked the loop.
+         * @param iterations the number iterations to report to the runtime system
+         * @since 0.12
+         */
+        protected abstract void onLoopCount(Node source, int iterations);
+
+        /** Makes sure the <code>callTarget</code> is initialized.
+         *
+         * @param callTarget
+         * @since 0.12
+         */
+        public void initializeCallTarget(RootCallTarget callTarget) {
+            ACCESSOR.initializeCallTarget(callTarget);
+        }
+
+        /** Finds the language associated with given root node.
+         *
+         * @param root the node
+         * @return the language of the node
+         * @since 0.12
+         */
+        @SuppressWarnings("rawtypes")
+        public Class<? extends TruffleLanguage> findLanguageClass(RootNode root) {
+            return ACCESSOR.findLanguage(root);
         }
     }
 
+    private static final class NoOp extends TVMCI {
+        private static final LoopNode NO_OP_NODE = new LoopNode() {
+            @Override
+            public void executeLoop(VirtualFrame frame) {
+                throw new UnsupportedOperationException();
+            }
+
+            @Override
+            public RepeatingNode getRepeatingNode() {
+                throw new UnsupportedOperationException();
+            }
+        };
+
+        NoOp() {
+            NO_OP_NODE.super();
+        }
```

Comment:
Good thing is that https://github.com/graalvm/truffle/pull/139/commits/d0ee89755dab0592ecc15769039a570daffdf4c8 removes this *object oriented fallback machinery* all together and replaces it with simple `if (SUPPORT == null)` check.